### ISSUE TITLE
Allow extensions commands to suppress CodeMirror6 default shortcuts

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#15897]
### Problem

As summarised in https://github.com/jupyterlab/jupyterlab/pull/14115#issuecomment-1974835451, with changes from listening to keydown events in capture phase to listening to them in bubbling phase (JupyterLab 4.1) we lost ability to `preventDefault()` the default CodeMirror shortcuts behaviour. The current solution is override respective CM commands manually and check if it should possibly be prevented (e.g. based on the class of a closest DOM node), but it does not work if user disables such command.

### Proposed Solution

Either:
1. Restore listening to events in the capture phase for _all_ shortcuts, or
2. Make it a property of lumino's IKeybinding so some shortcuts can be invoked already in capture phase and others only in bubbling phase, or
3. Allow extensions to require a plugin that would allow them to mark CM keybindings as ignored under specific circumstances, or
4. Modify CM6 keybinding implementation such that it invokes lumino's keyboard processor before proceeding with execution of CM6 command; basically have a pre-hook for all key interactions

### Additional context

- https://github.com/jupyterlab/jupyterlab/issues/15821
- https://github.com/jupyterlab-contrib/jupyterlab-vim/issues/138

**Comment by krassowski** [comment]
> but it does not work if user disables such a command or changes the shortcut.

Moreover, it makes extending jupyterlab difficult because one needs to hard-code these exceptions in core codemirror package to get the custom behaviour to work.

**Comment by gabalafou** [comment]
Can you give a specific example? If I open JupyterLab 4.4 and install the Vim extension and then go into notebook code cell, I can use Vim shortcuts to insert, append, cut, paste - but I'm don't use the Vim extension regularly, so I do not understand where the points of conflict between Vim and CodeMirror are in JupyterLab >= 4.1.

**Comment by daeyun** [comment]
I'd also appreciate an update on this. I'm stuck on an older version too. Certain default CodeMirror shortcuts, like Ctrl+K, can't be reassigned. Thanks for looking into this!

**Comment by gabalafou** [comment]
I see. So the problem is essentially the [table of Vim (Jupyter Edit mode) bindings](https://github.com/jupyterlab-contrib/jupyterlab-vim?tab=readme-ov-file#vim-jupyter-edit-mode-bindings).

**Comment by gabalafou** [comment]
I wonder if there's a fifth option not mentioned in this issue's description. It seems to me that CodeMirror 6 [does allow keys to be remapped](https://discuss.codemirror.net/t/overriding-certain-default-keymaps/6181) (also as evidenced by PR #15898), so maybe it would be possible for JupyterLab to expose this specific bit of CodeMirror to extension authors?

**Comment by daeyun** [comment]
> I see. So the problem is essentially the [table of Vim (Jupyter Edit mode) bindings](https://github.com/jupyterlab-contrib/jupyterlab-vim?tab=readme-ov-file#vim-jupyter-edit-mode-bindings).

Yes (and any custom bindings). Some of these shortcuts are commonly used and in muscle memory, so can be inconvenient to lose access to them.

It might be possible to hardcode a workaround, but it's tough to maintain across environments (e.g., Docker containers). For now, I'm downgrading since I don't need the latest JupyterLab features, even though that can be inconvenient when collaborating with others (as @Hnasar mentioned).

Happy to test if anyone's willing to update or fix. Thanks again!

**Comment by gabalafou** [comment]
Perhaps the most straightforward solution is to implement a mechanism for extension authors to configure keyboard shortcuts to be handled on the capture phase and hope that all of the various widgets and extensions in JupyterLab don't step on each other's toes too much. But I guess I'm not sure if that's a wise solution.

**Comment by daeyun** [comment]
> I see. So the problem is essentially the [table of Vim (Jupyter Edit mode) bindings](https://github.com/jupyterlab-contrib/jupyterlab-vim?tab=readme-ov-file#vim-jupyter-edit-mode-bindings).

Yes (and any custom bindings). Some of these shortcuts are commonly used and in muscle memory, so can be inconvenient to lose access to them.

It might be possible to hardcode a workaround, but it's tough to maintain across environments (e.g., Docker containers). For now, I'm downgrading since I don't need the latest JupyterLab features, but it can be inconvenient when collaborating with others (as @Hnasar mentioned).

Happy to help look into it or test if anyone's willing to update, though I'm not familiar with JupyterLab frontend. Thanks again!

**Comment by Hnasar** [comment]
> Can you give a specific example?
@gabalafou of course! I can give you two

Note this is running on a mac which has its own keyboard shortcuts: https://support.apple.com/en-us/102650#text

Example 1: We have black configured to run when editing a cell with _ctrl-b_. In newer jupyterlab, this doesn't run black, instead it runs the mac keyboard shortcut for 'move character left'. In the older jupyterlab, it properly runs the shortcut.

Example 2: With the vim plugin, ctrl-o, ctrl-o is supposed to insert a new cell below. In newer jupyterlab it does nothing (there's another macos shortcut). In older jupyterlab, it properly runs the shortcut.

Here is a demo recorded with https://github.com/wulkano/Kap
![Image](https://github.com/user-attachments/assets/837716e4-fb7a-4d9f-b652-84862d43915f)

As you might imagine, people are very tied to their shortcuts, so we tried to update… but decided to roll it back.

Thank you so much for taking a look!

**Comment by krassowski** [comment]
> I wonder if there's a fifth option not mentioned in this issue's description. It seems to me that CodeMirror 6 [does allow keys to be remapped](https://discuss.codemirror.net/t/overriding-certain-default-keymaps/6181) (also as evidenced by PR https://github.com/jupyterlab/jupyterlab/pull/15898), so maybe it would be possible for JupyterLab to expose this specific bit of CodeMirror to extension authors?

Which is option (3) in my original post (unless we want to expose raw code mirror internals). I don't necessarily think it is the best one though.

I think we should focus on option (2) or (4), with (4) having potential disadvantage of introducing slight overhead on typing, so preferably (2).

**Comment by gabalafou** [comment]
Yeah, I agree that option (2) is the most straightforward solution. Whoever implements this, I would kindly ask that you put a big warning sign in the code comments to other developers asking them to please not abuse the power of the capturing phase. 

If too many keyboard shortcuts get tied to the capturing phase, then we'll be right back to the situation where keyboard-only usage of JupyterLab is broken because certain keybindings that are critical for the keyboard operation of UI components no longer work because they are usurped at the capture phase by other modules or extensions.

**Comment by gabalafou** [comment]
For example, #14115—the escape key is critical for keyboard-only usage of CodeMirror, but it was previously being handled by JupyterLab in the capture phase, making it nearly impossible to navigate through notebooks using only the keyboard.

**Comment by krassowski** [comment]
> the escape key is critical for keyboard-only usage of CodeMirror, but it was previously being handled by JupyterLab in the capture phase

FYI, recent versions of CodeMirror have a built-in shortcut and a command for tab handling, see https://github.com/jupyterlab/jupyterlab/issues/17080

**Comment by gabalafou** [comment]
> [@gabalafou](https://github.com/gabalafou) of course! I can give you two

Thank you!

Just double checking, these problems with the keybindings only occur while in edit mode in the code cell, correct? (My assumption is that all of these conflicts are with CodeMirror's keybindings. And CodeMirror is instantiated only when the code cell is in edit mode.)

**Comment by gabalafou** [comment]
> FYI, recent versions of CodeMirror have a built-in shortcut and a command for tab handling, see [#17080](https://github.com/jupyterlab/jupyterlab/issues/17080)

Oh, interesting! Seems to me, though, that it would be preferable for the `escape` key to toggle out of "tab key indents" mode because for keyboard accessibility, the expectation is that in order to tab-navigate the page, the user shouldn't have to know any special keyboard shortcuts—should be able to get around using only escape, enter, tab, space, arrow keys, and nothing else.

**Comment by gabalafou** [comment]
Which is where it looks like #17080 ended up, so 👍

**Comment by firai** [comment]
What does it actually take to implement option 2? Partial reversal of https://github.com/jupyterlab/lumino/pull/653/files (and maybe of bits of https://github.com/jupyterlab/lumino/pull/635/files) to instead expose `bubblingKeyDown` in `IKeyBinding` in Lumino, with a default of `true`? (Should it be through `IKeyBinding` or `IKeyBindingOptions`, or both? I don't really understand how it works under the hood in Lumino. Are there more fundamental changes needed to make it work differently for some bindings versus others?)

The option then needs to be exposed through the interfaces in JupyterLab?

**Comment by firai** [comment]
After looking into this further, in order to implement option 2, it seems like another `addEventListener()` needs to be added around https://github.com/jupyterlab/lumino/blob/20cf366a88028ca763eb80241bd1226c936f454e/packages/application/src/index.ts#L359, so that one event listener has `useCapture` (the last argument) as `true` and the other has `false`, so that both capture and bubbling phases trigger the callback? Is that right?

And then something like https://github.com/jupyterlab/lumino/pull/689 needs to be added so that key bindings are processed based on whether a new `IKeyBinding` capture phase flag is true for a particular key binding? Is that right? Still haven't quite figured out how that PR works yet...

**Comment by daeyun** [comment]
If anyone needs a temporary workaround: here's a patch that disables all CodeMirror shortcuts by forcing `defaultKeymap` to be an empty list in the minified `.js` file. The filename varies across builds, so it has to search by content. If you only want to disable a specific shortcut, just search for the string (e.g., `"Ctrl-k"`) in the same file and remove its entry.

Tested with `jupyterlab==4.4.2` and `jupyterlab-vim==4.1.4`.
This will likely stop working in a future release if the internal structure or naming changes, but it works for now.

### Patch: Remove all CodeMirror shortcuts
```shell
find /usr/local/share/jupyter/lab/static -name '*.js' -exec perl -i -pe 's/(defaultKeymap:\(\)=>)([^,;}\[]{1,10})(?=[\W\[])/${1}[]\/\*$2\*\//' {} +
```

To locate your static directory:
```shell
echo "$(python3 -c 'from jupyterlab.labapp import LabApp; print(LabApp().app_dir)')/static"
# Use this value to replace /usr/local/share/jupyter/lab/static above.
```

### Dry run (preview only)
```shell
find /usr/local/share/jupyter/lab/static -name '*.js' -exec perl -ne 'if (/defaultKeymap:\(\)=>([^,;}\[]{1,10})(?=[\W\[])/) { print "$ARGV:\nBEFORE: defaultKeymap:()=>".$1."\nAFTER:  defaultKeymap:()=>[]/*$1*/\n" }' {} +
```
Expected output:
```
BEFORE: defaultKeymap:()=>wn
AFTER:  defaultKeymap:()=>[]/*wn*/
```

### Restore original shortcuts
```shell
# Dry run
find /usr/local/share/jupyter/lab/static -name '*.js' -exec perl -ne 'if (/defaultKeymap:\(\)=>\[\]\/\*([^*]+)\*\//) { print "$ARGV:\nBEFORE: defaultKeymap:()=>[]/*$1*/\nAFTER:  defaultKeymap:()=>$1\n" }' {} +

# Restore
find /usr/local/share/jupyter/lab/static -name '*.js' -exec perl -i -pe 's/(defaultKeymap:\(\)=>)\[\]\/\*([^*]*)\*\//${1}$2/' {} +
```
Expected dry run output:
```
BEFORE: defaultKeymap:()=>[]/*wn*/
AFTER:  defaultKeymap:()=>wn
```

### Dockerfile example

As long as you use fixed versions, it won't break:
```dockerfile
RUN pip install -U jupyterlab==4.4.2 jupyterlab-vim==4.1.4 && \
    JUPYTERLAB_STATIC=$(python3 -c 'from jupyterlab.labapp import LabApp; print(LabApp().app_dir + "/static")') && \
    find "$JUPYTERLAB_STATIC" -name '*.js' -exec perl -i -pe 's/(defaultKeymap:\(\)=>)([^,;}\[]{1,10})(?=[\W\[])/${1}[]\/\*$2\*\//' {} +
```

**Comment by daeyun** [comment]
If anyone needs a temporary workaround: Resetting `defaultKeymap` in one of the minified `.js` files disables all CodeMirror shortcuts. Tested with `jupyterlab==4.4.2` and `jupyterlab-vim==4.1.4`.
This will likely stop working in a future release if the internal structure or naming changes.

Remove CodeMirror shortcuts:
```shell
find /usr/local/share/jupyter/lab/static -name '*.js' -exec perl -i -pe 's/(defaultKeymap:\(\)=>)([^,;}\[]{1,10})(?=[\W\[])/${1}[]\/\*$2\*\//' {} +
```

To locate your static directory:
```shell
echo "$(python3 -c 'from jupyterlab.labapp import LabApp; print(LabApp().app_dir)')/static"
# Use this value to replace /usr/local/share/jupyter/lab/static above.
```

Dry run (preview only):
```shell
find /usr/local/share/jupyter/lab/static -name '*.js' -exec perl -ne 'if (/defaultKeymap:\(\)=>([^,;}\[]{1,10})(?=[\W\[])/) { print "$ARGV:\nBEFORE: defaultKeymap:()=>".$1."\nAFTER:  defaultKeymap:()=>[]/*$1*/\n" }' {} +
```
Expected output:
```
/usr/local/share/jupyter/lab/static/4353.8572f6845cfde92dc152.js:
BEFORE: defaultKeymap:()=>wn
AFTER:  defaultKeymap:()=>[]/*wn*/
```

Restore original shortcuts
```shell
# Dry run
find /usr/local/share/jupyter/lab/static -name '*.js' -exec perl -ne 'if (/defaultKeymap:\(\)=>\[\]\/\*([^*]+)\*\//) { print "$ARGV:\nBEFORE: defaultKeymap:()=>[]/*$1*/\nAFTER:  defaultKeymap:()=>$1\n" }' {} +

# Restore
find /usr/local/share/jupyter/lab/static -name '*.js' -exec perl -i -pe 's/(defaultKeymap:\(\)=>)\[\]\/\*([^*]*)\*\//${1}$2/' {} +
```
Expected dry run output:
```
/usr/local/share/jupyter/lab/static/4353.8572f6845cfde92dc152.js:
BEFORE: defaultKeymap:()=>[]/*wn*/
AFTER:  defaultKeymap:()=>wn
```

Dockerfile:

As long as you use fixed versions, it won't break:
```dockerfile
RUN pip install -U jupyterlab==4.4.2 jupyterlab-vim==4.1.4 && \
    JUPYTERLAB_STATIC=$(python3 -c 'from jupyterlab.labapp import LabApp; print(LabApp().app_dir + "/static")') && \
    find "$JUPYTERLAB_STATIC" -name '*.js' -exec perl -i -pe 's/(defaultKeymap:\(\)=>)([^,;}\[]{1,10})(?=[\W\[])/${1}[]\/\*$2\*\//' {} +
```

**Comment by firai** [comment]
@gabalafou @krassowski Would it be possible for you to give more pointers on how option 2 could be implemented? Is what I said above in line with what you're thinking, or do you envision accomplishing this differently? Any help would be greatly appreciated!

Should an issue also be filed at the lumino repo?

**Comment by krassowski** [comment]
>  Is what I said above in line with what you're thinking, or do you envision accomplishing this differently

I do not have a detailed implementation plain in my head right now, but what you proposed in https://github.com/jupyterlab/jupyterlab/issues/15897#issuecomment-2867535580 sounds about right.

> Should an issue also be filed at the lumino repo?

Feel welcome to open an issue on lumino, it is always nice to keep other lumino users in loop about the plans.
